File: Cpp/_102.cpp
/*
Use dummy elem (None) to indicate end of current
level, then add dummy for next level. Level-order 
requires breadth-first, which requires queue
*/

// Iterative
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return vector<vector<int>>();
        queue<TreeNode*> q;
        q.push(root);
        q.push(nullptr);
        
        vector<vector<int>> li;
        vector<int> levelLi;
        
        while (!q.empty()) {
            TreeNode* current = q.front();
            q.pop();
            
            if (current) {
                levelLi.push_back(current->val);
                if (current->left) q.push(current->left);
                if (current->right) q.push(current->right);
            }
            else {
                li.push_back(levelLi);
                levelLi.clear();
                if (!q.empty()) q.push(nullptr);
            }
        }
        
        return li;
    }
};

File: Python3/_102.py
"""
Use dummy elem (None) to indicate end of current
level, then add dummy for next level. Level-order 
requires breadth-first, which requires queue
"""

class Solution:
    def levelOrder(self, root):
        if not root:
            return []
        
        li, levelLi, q = [], [], deque([root, None])
        while q:
            current = q.popleft()

            # End of current level, prepare for next
            # level if there are remaining nodes
            if not current:
                if q:
                    q.append(None)
                    
                li.append(levelLi)
                levelLi = []
                continue
                
            levelLi.append(current.val)
            if current.left:
                q.append(current.left)
            if current.right:
                q.append(current.right)
        
        return li